### PR TITLE
[SHELL32] Add 'Create New Extension' IDD_NEWEXTENSION dialog

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-12906](https://jira.reactos.org/browse/CORE-12906)

This PR will add the 'Create New Extension' dialog (IDD_NEWEXTENSION).

## Proposed changes

- Add IDS_NEWEXT_ADVANCED_LEFT, IDS_NEWEXT_ADVANCED_RIGHT, IDS_NEWEXT_NEW, IDS_NEWEXT_SPECIFY_EXT, IDS_NEWEXT_ALREADY_ASSOC, and IDS_NEWEXT_EXT_IN_USE resource strings.
- Add IDD_NEWEXTENSION resource dialog.
- Add the NewExtensionDlgProc dialog procedure.
- Enable the user to add a new extension.

### CODE REVIEW

**Comment by katahiromz** [comment]
2k3:
2k3-newext
2k3-newext-expanded

ros:
ros-newext
ros-newext-expanded

**Comment by katahiromz** [comment]
ros:
ros-newext-2
ros-newext-3-expanded

**Comment by katahiromz** [comment]
ros:
ros-newext-4
ros-newext-4-expanded

**Comment by katahiromz** [comment]
ros-test-2
Done. Please test and commit me.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 1]
```diff
@@ -1556,6 +1556,311 @@ FindSelectedItem(
...
+    pszText[cch] = 0;
+}
+
+static BOOL
+NewExtDlg_OnOK(HWND hwndDlg, NEWEXT_DIALOG *pNewExt)
+{
+    LV_FINDINFO find;
+    INT iItem;
+
+    GetDlgItemText(hwndDlg, IDC_NEWEXT_EDIT, pNewExt->szExt, _countof(pNewExt->szExt));
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/folder_options.cpp]: GetDlgItemTextW.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 2]
```diff
@@ -1556,6 +1556,311 @@ FindSelectedItem(
...
+    StringTrimW(pNewExt->szExt);
+    CharUpperW(pNewExt->szExt);
+
+    GetDlgItemText(hwndDlg, IDC_NEWEXT_COMBOBOX, pNewExt->szFileType, _countof(pNewExt->szFileType));
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/folder_options.cpp]: GetDlgItemTextW.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 3]
```diff
@@ -1556,6 +1556,311 @@ FindSelectedItem(
...
+
+static BOOL
+FileTypesDlg_AddExt(HWND hwndDlg, LPCWSTR pszExt, LPCWSTR pszFileType)
+{
+    DWORD dwValue = 1;
+    HKEY hKey;
+    WCHAR szKey[MAX_PATH];
+    LONG nResult;
+
+    // Search the next "ft%06u" key name
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/folder_options.cpp]: So this is also what Windows does, i.e. it uses by default a "ft..." keyname? Also, only 6 digits?

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 4]
```diff
@@ -1556,6 +1556,311 @@ FindSelectedItem(
...
+    {
+        StringCchPrintf(szKey, _countof(szKey), TEXT("ft%06u"), dwValue);
+
+        nResult = RegOpenKeyEx(HKEY_CLASSES_ROOT, szKey, 0, KEY_READ, &hKey);
+        if (nResult != ERROR_SUCCESS)
+            break;
+
+        RegCloseKey(hKey);
+        ++dwValue;
+    } while (dwValue < 0x1000000);
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/folder_options.cpp]: Is that value "0x1000000" arbitrary, or can it go up to 0xFFFFFFFF ?

Also, note that: ft%06u means you're going to insert 6 digits after "ft" (possibly 0-padded), but 0x1000000 == 16777216 in decimal, which is longer than 6 digits.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 5]
```diff
@@ -1556,6 +1556,311 @@ FindSelectedItem(
...
+    if (*pszExt == L'.')
+        ++pszExt;
+    StringCchPrintf(szExt, _countof(szExt), TEXT(".%s"), pszExt);
+    CharLowerW(szExt);
+    nResult = RegCreateKeyEx(HKEY_CLASSES_ROOT, szExt, 0, NULL, 0, KEY_WRITE, NULL, &hKey, NULL);
+    CharUpperW(szExt);
+    if (ERROR_SUCCESS != nResult)
+        return FALSE;
+
+    // ".ext" @ --> "ft%06u"
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/folder_options.cpp]: Can you write a sentence?

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 6]
```diff
@@ -1578,14 +1884,22 @@ FolderOptionsFileTypesDlg(
...
             return TRUE;
 
         case WM_COMMAND:
             switch(LOWORD(wParam))
             {
-                case 14006:
-                    pItem = FindSelectedItem(GetDlgItem(hwndDlg, 14000));
+                case IDC_FILETYPES_NEW:
+                    newext.hwndLV = GetDlgItem(hwndDlg, IDC_FILETYPES_LISTVIEW);
+                    if (IDOK == DialogBoxParam(shell32_hInstance, MAKEINTRESOURCE(IDD_NEWEXTENSION),
```
Comment (HBelusca) [inline dll/win32/shell32/dialogs/folder_options.cpp]: DialogBoxParamW and MAKEINTRESOURCEW .

**Comment by HBelusca** [comment]
There are few things to check, but otherwise the patch looks good :D

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: OK.

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: OK.

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 6 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: OK.

File: dll/win32/shell32/dialogs/folder_options.cpp [hunk 7]
```diff
@@ -1556,6 +1556,311 @@ FindSelectedItem(
...
+        LoadStringW(shell32_hInstance, IDS_NEWEXT_ADVANCED_RIGHT, szText, _countof(szText));
+        SetDlgItemTextW(hwndDlg, IDC_NEWEXT_ADVANCED, szText);
+
+        rc.bottom -= pNewExt->dy;
+
+        LoadStringW(shell32_hInstance, IDS_NEWEXT_NEW, szText, _countof(szText));
+        SetDlgItemTextW(hwndDlg, IDC_NEWEXT_COMBOBOX, szText);
+    }
+
+    MoveWindow(GetDlgItem(hwndDlg, IDOK), rc1.left, rc1.top, rc1.right - rc1.left, rc1.bottom - rc1.top, TRUE);
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: OK.

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: decimal
Digits are decimal. And max 13 characters ("ft4294967295" + "\0"). OK.

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: OK.

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: // Set the default value of ".ext" to "ft%06u"

OK.

**Comment by katahiromz** [comment]
Done.

File: dll/win32/shell32/dialogs/folder_options.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/dialogs/folder_options.cpp]: I thought it's better to use 0xFFFFFFFF for the upper bound. Thank you.

Comment (HBelusca) [reply dll/win32/shell32/dialogs/folder_options.cpp]: :D Then possibly use the "ULONG_MAX" macro (that is equal to 0xFFFFFFFF).

**Pull request merged**